### DISCUSSION TITLE
How to extract the relationship between the genes inside a biological pathway diagram?

### DISCUSSION THREAD

**Discussion by arulmabr** [wikipathways/wikipathways-help#131] [Q&A] [answered]
I’m working on a project to extract the following details from a set of biological diagrams (all of them were obtained from wikipathways): Source 1 (tail of the arrow), Source 2 (head of the arrow), if possible their relationship like activates, inhibits etc.  

From my search so far I couldn’t find any data for the how the genes inside a figures are connected/related. Is there any database or a technique to extract these data? The most important detail necessary is the interlinking of arrows and preferably the relationship.  

For example:
https://pfocr.wikipathways.org/figures/PMC1635984__F3 

![1471-2164-7-279-3](https://github.com/user-attachments/assets/1ef5714b-ddd2-44b8-87aa-bae8840a72a4)

  Extracted data through ChatGPT (it’s not perfect, misses out a lot of details):

<img width="294" alt="Screenshot 2024-09-09 at 11 25 51 AM" src="https://github.com/user-attachments/assets/9f713592-59da-4124-b04a-e9de86ea6634">

Please tell me if there is a better to way to find these information or any suggestions/advise welcome :)
Thanks a lot.

**Comment by khanspers** [comment, marked as answer]
Hi @arulmabr. The link you provided is from [Pathway Figure OCR](https://pfocr.wikipathways.org/) (PFOCR), which is separate from [WikiPathways](https://www.wikipathways.org/). The PFOCR project extracts pathway information from published pathway figures, whereas WikiPathways is a database for creating, contributing and sharing pathway models. 

PFOCR doesn't yet contain information on interactions, but @AlexanderPico has developed a GPT to extract additional information, including interactions. You can find it here: https://chatgpt.com/g/g-VkkN5JFiG-pathway-figure-reader.

**Reply by arulmabr** [reply]
Hi @khanspers,
Can you please help me with getting the interactions from WikiPathways links like https://www.wikipathways.org/pathways/WP3271.html if possible? This data looks well organized than PFOCR ones for parsing, from the looks of it. Hopefully there is a way. Thanks!

**Reply by khanspers** [reply]
Hi. For WikiPathways, you can get interactions for a specific pathway from the RDF: https://sparql.wikipathways.org/

On the right side, click on **D. General**, then **InteractionsofPathway.rq**. Swap out the WPID in the query for WP3271.

**Reply by arulmabr** [reply]
Thank you @khanspers 
When I query that I get these results 
<img width="1425" alt="image" src="https://github.com/user-attachments/assets/dfd85db9-f1f7-4e8e-8bbd-61b13d68f9da">
Can you please tell me how to access more details here, as the link under the interaction columns aren't working.
For ex: http://rdf.wikipathways.org/Pathway/WP1425_r126273/ComplexBinding/a9b99

Once again, thanks for your help!

**Reply by khanspers** [reply]
I'm not sure why that link is broken. Another option is to import the pathway to Cytoscape using the [WikiPathways App](https://apps.cytoscape.org/apps/wikipathways) and then exporting the **Edge Table** to a text file via File > Export > Table to File.

**Reply by arulmabr** [reply]
Thanks a lot @khanspers!!
Appreciate the help :)

**Comment by arulmabr** [comment]
Hey @khanspers, thanks a lot for getting back and for the clarification!!
Just checked out @AlexanderPico's GPT, it's really good.
For our research, we need to process in bulk so API is more suitable. I'll reach out to @AlexanderPico for more help on this. 
Thank you :)

**Comment by AlexanderPico** [comment]
Hi. Sorry I'm late to this discussion.  A couple points that may be helpful...

1. PFOCR only extracts entities (genes, chemicals and diseases)
2. It is currently not possible to extract interactions. PFOCR doesn't do it. No tool nor algorithm I'm aware of can do it. This is an active area of research that I suspect is still at least 2-3 years away from being available.
3. The Pathway Reader GPT is just a quick toy I threw together to test the current "off the shelf" capabilities of LLMs to extract interactions. It is really bad at it! Do not trust these results! As you saw, it gets some right, but misses many others and hallucinates many as well.
4. Only WikiPathways and other human-curated pathway databases (e.g., Reactome and GO-CAMs) capture interactions with mechanistic detail (e.g., activates, inhibits, etc).  Interaction databases like STRING-DB and IntAct capture a broader set of relationships between proteins, but without mechanistic detail, e.g., A co-precipates with B, or A co-occurs in the literature with B, or A is often co-expressed with B.
5. RDF is the best way to get the most accurate interaction information from WikiPathways. Note: RDF uses URLs (called URIs in this context) to represent each entity and interaction. These links are not supposed to be "clickable" leading to a website, per se. They are just a standard way to produce unique identifiers.
6. Using the Edge Table from Cytoscape is good (though inferior to RDF). You will want to use import-as-network to at least represent interactions with complexes and hyperedges properly.  Regardless, I wouldn't trust these 100% for computational inference, for example. But I don't know your use case.

**Comment by arulmabr** [comment]
Thank you, @AlexanderPico, for your input!

Regarding the items in WikiPathways, could you provide an example of how to retrieve interaction information using RDF? Here’s my use case: I am only interested in identifying pairs of genes that are directly connected in a pathway, regardless of the direction or type of interaction. Specifically, if a pathway is represented as A --> B --> C, I would like to generate a dataset containing the pairs (A, B) and (B, C), while excluding indirect pairs like (A, C).

I tried the SPARQL for [WP4589](https://www.wikipathways.org/pathways/WP4589.html) and in the [queried results](https://sparql.wikipathways.org/?q=SELECT%20DISTINCT%20%3Fpathway%20%3Finteraction%20%3Fparticipants%20%3FDataNodeLabel%0AWHERE%20%7B%0A%0A%20%20%20%3Fpathway%20a%20wp%3APathway%20%3B%0A%20%20%20%20%20%20dc%3Aidentifier%20%3Chttps%3A%2F%2Fidentifiers.org%2Fwikipathways%2FWP4589%3E%20.%0A%20%20%20%3Finteraction%20dcterms%3AisPartOf%20%3Fpathway%20%3B%20%0A%20%20%20%20%20%20a%20wp%3AInteraction%20%3B%0A%20%20%20%20%20%20wp%3Aparticipants%20%3Fparticipants%20.%0A%20%20%20%3Fparticipants%20a%20wp%3ADataNode%20%3B%0A%20%20%20%20%20%20rdfs%3Alabel%20%3FDataNodeLabel%20.%20%20%0A%7D%0A), I couldn't find the entities Threonine, Serine etc.

I tried exporting from Cytoscape as an edge table and there as well these entities are classified as null. Then I manually went through the GPML file and noticed that the structuring is lacking for these elements. The corresponding GraphId for these elements from their <DataNode> aren’t under any of the <Interaction> data. 

Is there any other way to capture all these, or am I missing something here. Please help me out. Thanks in advance!
[WP4589.zip](https://github.com/user-attachments/files/17303454/WP4589.zip)

**Comment by AlexanderPico** [comment]
Looking at WP4589, I don't see Serine or Threonine connected to anything. They appear to be in a complex labeled "A protein" which is connected to things. 

So, neither Cytoscape, nor RDF, nor any format will give you connections between those amino acids and anything else because they are not connected to anything else.

What type of interaction (e.g., Serine --> ??) would you expect to get from this pathway?

**Reply by arulmabr** [reply]
Okay, understood.

So Ideally we want to capture all the interactions like:
Serine --> Glycogen Breakdown,
Serine --> Glycogen Synthesis,
Serine --> Glycolysis etc.

**Comment by AlexanderPico** [comment]
Got it. That makes sense. Unfortunately, at this time, the only technology capable of capturing those interactions from this figure is... the human brain :)  And only the small subset of brains that are trained in Biology (like you) that can understand the meaning of this drawing.

This is an open challenge in the field of pathway modeling: How to draw pathways in a way that are understood by BOTH humans and computers. Projects like SBML have focused on the computer interpretation and require strict rules when drawing a pathway in order to ensure computation analysis of every detail. Most pathway diagrams (~1000 published each month) do not follow ANY modeling standards and are drawn in PowerPoint or Illustrator.  With WikiPathways, we try to find a compromise where we require some things (like node xrefs and connected interactions), but we also let people draw things that will be intuitive for humans even if they are not for computers.

So, you can get SOME interactions from RDF and Cytoscape edge table, but you cannot get ALL the interactions that you (as a human) can understand from every pathway drawing.

**Reply by arulmabr** [reply]
Agreed, thanks a lot for explaining these and helping me out.
Really appreciate it @AlexanderPico :)
